### DISCUSSION TITLE
Question: Semaphore Uses Wrong Ansible Binary

### DISCUSSION THREAD

**Discussion by kupe-dono** [semaphoreui/semaphore#2654] [Q&A] [answered]
### Question

Hello there!

I got some issues lately that I solved but can't figure out why it occurs.
When I run a playbook from Semaphore UI I got this:

```
2:24:10 PM
ERROR! Unexpected Exception, this is probably a bug: module 'ansible.constants' has no attribute 'CONTROLLER_PYTHON_WARNING'
2:24:10 PM
the full traceback was:
2:24:10 PM
2:24:10 PM
Traceback (most recent call last):
2:24:10 PM
  File "/usr/local/bin/ansible-playbook", line 77, in <module>
2:24:10 PM
    if C.CONTROLLER_PYTHON_WARNING and not _PY38_MIN:
2:24:10 PM
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2:24:10 PM
AttributeError: module 'ansible.constants' has no attribute 'CONTROLLER_PYTHON_WARNING'

```
It appears that Semaphore uses binaries from "/usr/local/bin" instead of "/usr/bin/".

But the Ansible version installed is well configured to use /usr/bin binaries and works fine in CLI:

```
ansible [core 2.14.16]
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /bin/ansible
  python version = 3.11.2 (main, Sep 14 2024, 03:00:30) [GCC 12.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True

```
My temporary fix > create a symbolic link to force semaphore using the right binaries.

**Is someone has a explanation about why the UI is using the binaries in /usr/local/bin?**

My Config:

- Debian 12
- Ansible last version
- Python 3.11
- Semaphore 2.10.43 installed with dpkg

**Comment by fiftin** [comment, marked as answer]
@kupe-dono, this isn’t an issue with Semaphore but rather with your environment. 

I suspect Ansible is installed in two different locations on your system, and they appear in a different order within your `PATH` environment variable.

You can specify `PATH` environment variable for Semaphore explicitly like in your user's environment.

**Reply by kupe-dono** [reply]
Thank you very much @fiftin I will look into it, I was wondering if Semaphore was installing its own ansible binaries in /usr/local/bin with dpkg but it seems not. So you're probably right :) the issue is on the server side and not from Semaphore!

**Reply by fiftin** [reply]
We planning to add Ansible version management to Semaphore. But it will be VENVs, not your system ;)

**Reply by PeterFalken** [reply]
Thanks @fiftin, for planning on adding Ansible Version Management to Semaphore, I lost the ability to use SemaphoreUI with older machines when I upgraded from `2.10.43` to `2.11.1`, `ansible-core` 2.17+ or later do not support python versions older than 3.7+ on the target hosts, reference [here](https://github.com/ansible/ansible/issues/82068).

**Reply by mlanner** [reply]
@fiftin I'm very interested in the feature you mention above, about Ansible version management, and we did have a conversation about it in issue https://github.com/semaphoreui/semaphore/issues/2682 too. But, a few questions on that as referenced above:

* You say it will be VENVs. Would it be possible to call containers instead of Python/pip VENVs? I feel like it would maybe be more "stable" and repeatable to build it in containers, having Dockerfiles that one can modify/share with the community.
* Would it be possible to have multiple VENVs/containers and be able to call them specifically for some tasks that require a certain version of Ansible? For example, for the vast majority of day-to-day tasks, using Ansible 2.18 works fine for me/us, but for some legacy systems (Rocky Linux 8), we need to use Ansible 2.16. I would prefer to not have to run two different Semaphore instances for this, as that gets confusing for users.

Sorry for the cross-post here, but when I did a little searching in the project I found this discussion and wanted to also draw the participants attention to the issue and discussion there. Plus, if anyone has come up with some kind of intermittent solution for this, I'm very interested in any ideas that could help with the problem I'm currently up against.

**Reply by PeterFalken** [reply]
@mlanner, I agree with you that running two instances of Semaphore can be cumbersome. There might be a way to run two containers as `runner nodes` and dispatch tasks to them according to the inventory that is targeted - I think that this was feature introduced on [version 2.13](https://semaphoreui.com/releases/semaphore-v2_13)

I would prefer that this feature be implemented as `VENVs`, and that you could choose on the task template definition which VENV you would like them to be run on. By default new task template run on the default VENV that uses the `ansible` version shipped with Semaphore UI, but if you need to, you can set a different `ansible` version on the Task Template UI, this would create a `VENV` under `/opt/semaphore/apps/ansible` that would be used to run any task that uses that particular `ansible` version. This way VENVs can be reused for multiple tasks, while running only one instance of Semaphore; which also simplifies deployments.

**Reply by mlanner** [reply]
@PeterFalken Thanks for the quick response and ideas. I've experimented with VENVs, mainly building them using UV. That sort of worked fine, except that I couldn't target the different VENVs that I installed and configured through Semaphore. I could switch between them running the Ansible CLI, of course, but not with Semaphore. In my installation I don't have the Runners feature and frankly I hadn't looked into it much in terms of what I thought I wanted/needed. I'll look into it a bit.

Also, I have Semaphore installed from 'native' packages. So, I'm not using the container deployment method, which bundles the various other components with it. To use the container deployment, I'd prefer to see each piece (Semaphore, Ansible, Terraform/OpenTofu, etc.) implemented as separate containers, which could then be versioned individually. That would also potentially allow me to run more than one Ansible container and therefore target a specific Ansible version that way. AFAIK, that's not really a possibility today. But I may have missed something, and if so, I'd be delighted to be corrected. :smiley:

**Reply by PeterFalken** [reply]
You might want to take a look at these Dockerfiles:
[server](https://github.com/semaphoreui/semaphore/blob/develop/deployment/docker/server/Dockerfile)
[runner](https://github.com/semaphoreui/semaphore/blob/develop/deployment/docker/runner/Dockerfile)
Where you can set the versions for the `pieces/tools`.
